sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubClient.java

```diff
@@ -0,0 +1,493 @@
+      }
+
+      OutgoingMessage that = (OutgoingMessage) o;
+
+      if (timestampMsSinceEpoch != that.timestampMsSinceEpoch) {
+        return false;
+      }
+      return Arrays.equals(elementBytes, that.elementBytes);
+
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(Arrays.hashCode(elementBytes), timestampMsSinceEpoch);
+    }
+  }
+
+  /**
+   * A message received from Pubsub.
+   */
+  public static class IncomingMessage {
+    /**
+     * Underlying (encoded) element.
+     */
+    public final byte[] elementBytes;
+
+    /**
+     * Timestamp for element (ms since epoch). Either Pubsub's processing time,
+     * or the custom timestamp associated with the message.
+     */
+    public final long timestampMsSinceEpoch;
+
+    /**
+     * Timestamp (in system time) at which we requested the message (ms since epoch).
+     */
+    public final long requestTimeMsSinceEpoch;
+
+    /**
+     * Id to pass back to Pubsub to acknowledge receipt of this message.
+     */
+    public final String ackId;
+
+    /**
+     * Id to pass to the runner to distinguish this message from all others.
+     */
+    public final byte[] recordId;
+
+    public IncomingMessage(
+        byte[] elementBytes,
+        long timestampMsSinceEpoch,
+        long requestTimeMsSinceEpoch,
+        String ackId,
+        byte[] recordId) {
+      this.elementBytes = elementBytes;
+      this.timestampMsSinceEpoch = timestampMsSinceEpoch;
+      this.requestTimeMsSinceEpoch = requestTimeMsSinceEpoch;
+      this.ackId = ackId;
+      this.recordId = recordId;
+    }
+
+    public IncomingMessage withRequestTime(long requestTimeMsSinceEpoch) {
+      return new IncomingMessage(elementBytes, timestampMsSinceEpoch, requestTimeMsSinceEpoch,
+                                 ackId, recordId);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      IncomingMessage that = (IncomingMessage) o;
+
+      if (timestampMsSinceEpoch != that.timestampMsSinceEpoch) {
+        return false;
+      }
+      if (requestTimeMsSinceEpoch != that.requestTimeMsSinceEpoch) {
+        return false;
+      }
+      if (!Arrays.equals(elementBytes, that.elementBytes)) {
+        return false;
+      }
+      if (!ackId.equals(that.ackId)) {
+        return false;
+      }
+      return Arrays.equals(recordId, that.recordId);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(Arrays.hashCode(elementBytes), timestampMsSinceEpoch,
+                              requestTimeMsSinceEpoch,
+                              ackId, Arrays.hashCode(recordId));
+    }
+  }
+
+  /**
+   * @inheritDoc
+   */
+  @Override
+  public abstract void close() throws IOException;
+
+  /**
+   * Publish {@code outgoingMessages} to Pubsub {@code topic}. Return number of messages
+   * published.
```

Comment:
(github won't let me chain off the previous discussion).

So what purpose does the return value serve? Should this function just be void?